File: apps/mobile/lib/sync.ts
import * as SecureStore from "expo-secure-store";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { AppState, type AppStateStatus } from "react-native";
import { getAllTasksForSync, batchUpsertFromSync, getAllTasks } from "./tasks";
import { rescheduleAllAlarms } from "./alarm";
import type { SyncTask } from "@cluanote/shared";

export const SYNC_URL_STORE_KEY = "cluanote_sync_url";
export const AUTO_SYNC_ENABLED_KEY = "cluanote_auto_sync_enabled";
export const LAST_SYNCED_AT_KEY = "cluanote_last_synced_at";

export interface SyncConfig {
  isConfigured: boolean;
  url: string | null;
  autoSync: boolean;
  lastSyncedAt: string | null;
}

export interface SyncResult {
  success: boolean;
  pushedCount: number;
  pulledCount: number;
  pulledTasks?: SyncTask[];
  syncedAt: string;
  message: string;
}

/**
 * Retrieves the stored synchronization configuration.
 */
export async function getSyncConfig(): Promise<SyncConfig> {
  try {
    let url: string | null = null;
    try {
      url = await SecureStore.getItemAsync(SYNC_URL_STORE_KEY);
    } catch {
      url = await AsyncStorage.getItem(SYNC_URL_STORE_KEY);
    }

    const autoSyncRaw = await AsyncStorage.getItem(AUTO_SYNC_ENABLED_KEY);
    const lastSyncedAt = await AsyncStorage.getItem(LAST_SYNCED_AT_KEY);
    const autoSync = autoSyncRaw === null ? true : autoSyncRaw === "true";

    return {
      isConfigured: Boolean(url && url.trim().length > 0),
      url,
      autoSync,
      lastSyncedAt,
    };
  } catch (err) {
    console.warn("Failed to read sync config:", err);
    return {
      isConfigured: false,
      url: null,
      autoSync: false,
      lastSyncedAt: null,
    };
  }
}

/**
 * Saves and securely persists the synchronization URL and auto-sync preference.
 */
export async function saveSyncConfig(
  url: string,
  autoSync = true
): Promise<void> {
  const trimmed = url.trim();
  try {
    await SecureStore.setItemAsync(SYNC_URL_STORE_KEY, trimmed);
  } catch {
    await AsyncStorage.setItem(SYNC_URL_STORE_KEY, trimmed);
  }
  await AsyncStorage.setItem(AUTO_SYNC_ENABLED_KEY, String(autoSync));
}

/**
 * Removes the sync URL and turns off sync.
 */
export async function disconnectSync(): Promise<void> {
  try {
    await SecureStore.deleteItemAsync(SYNC_URL_STORE_KEY);
  } catch {
    // Ignore
  }
  await AsyncStorage.removeItem(SYNC_URL_STORE_KEY);
  await AsyncStorage.setItem(AUTO_SYNC_ENABLED_KEY, "false");
}

/**
 * Helper to determine if a URL is a Neon PostgreSQL connection string.
 */
function isNeonConnectionString(raw: string): boolean {
  return (
    (raw.startsWith("postgres://") || raw.startsWith("postgresql://")) &&
    raw.includes("neon.tech")
  );
}

/**
 * Executes a SQL query against Neon PostgreSQL over secure HTTPS.
 */
async function executeNeonSql(
  pgUrl: string,
  query: string,
  params: unknown[] = []
): Promise<Record<string, unknown>[]> {
  const urlObj = new URL(pgUrl);
  const host = urlObj.hostname;
  const password = decodeURIComponent(urlObj.password);

  const endpoint = `https://${host}/sql`;
  const response = await fetch(endpoint, {
    method: "POST",
    headers: {
      "Authorization": `Bearer ${password}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      query,
      params,
    }),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Neon SQL error (${response.status}): ${errorText}`);
  }

  const data = await response.json();
  return (data.rows || []) as Record<string, unknown>[];
}

/**
 * Tests connection to the remote PostgreSQL / Neon database or sync server.
 */
export async function testSyncConnection(url: string): Promise<{ success: boolean; message: string }> {
  const trimmed = url.trim();
  if (!trimmed) {
    throw new Error("Connection URL cannot be empty");
  }

  if (isNeonConnectionString(trimmed)) {
    // Direct Neon HTTPS test
    try {
      const rows = await executeNeonSql(trimmed, "SELECT 1 as connected, NOW() as server_time;");
      if (rows && rows.length > 0) {
        return {
          success: true,
          message: `Connected to Neon PostgreSQL successfully! Server time: ${rows[0].server_time || "OK"}`,
        };
      }
      return { success: true, message: "Connected successfully." };
    } catch (err) {
      throw new Error(`Connection failed: ${err instanceof Error ? err.message : String(err)}`);
    }
  }

  // Generic Sync Server HTTP test
  try {
    const cleanUrl = trimmed.replace(/\/+$/, "");
    const testUrl = cleanUrl.endsWith("/api/test") ? cleanUrl : `${cleanUrl}/api/test`;
    const response = await fetch(testUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ url: trimmed }),
    });

    if (!response.ok) {
      const txt = await response.text();
      throw new Error(`Server returned HTTP ${response.status}: ${txt}`);
    }

    const json = await response.json();
    return {
      success: true,
      message: json.message || "Connected to sync server successfully!",
    };
  } catch (err) {
    throw new Error(`Failed to reach sync server: ${err instanceof Error ? err.message : String(err)}`);
  }
}

/**
 * Runs a complete bi-directional sync:
 * 1. Fetches latest remote tasks FIRST.
 * 2. Merges any newly added or updated remote tasks into local SQLite.
 * 3. Selectively pushes only newer local tasks to remote.
 * 4. Reschedules upcoming active alarms.
 */
export async function runPostgresSync(): Promise<SyncResult> {
  const config = await getSyncConfig();
  if (!config.url) {
    throw new Error("No synchronization URL configured. Please configure your database in Settings.");
  }

  const url = config.url.trim();
  const localTasks = await getAllTasksForSync();
  const localMap = new Map<string, SyncTask>(localTasks.map((t) => [t.uuid, t]));

  let remoteTasks: SyncTask[] = [];

  if (isNeonConnectionString(url)) {
    // 1. Ensure table schema exists on Neon PostgreSQL
    await executeNeonSql(
      url,
      `
      CREATE TABLE IF NOT EXISTS cluanote_tasks (
          uuid VARCHAR(64) PRIMARY KEY,
          title TEXT NOT NULL,
          note TEXT,
          date VARCHAR(10) NOT NULL,
          time VARCHAR(10),
          priority VARCHAR(10) NOT NULL DEFAULT 'medium',
          completed INTEGER NOT NULL DEFAULT 0,
          created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
          updated_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
          is_deleted BOOLEAN NOT NULL DEFAULT FALSE,
          is_future_note INTEGER NOT NULL DEFAULT 0
      );
      CREATE INDEX IF NOT EXISTS idx_cluanote_tasks_date ON cluanote_tasks(date);
      CREATE INDEX IF NOT EXISTS idx_cluanote_tasks_updated ON cluanote_tasks(updated_at);
      ALTER TABLE cluanote_tasks ADD COLUMN IF NOT EXISTS is_future_note INTEGER NOT NULL DEFAULT 0;
      `
    );

    // 2. Fetch all remote tasks FIRST
    const rows = await executeNeonSql(
      url,
      `
      SELECT 
          uuid, 
          title, 
          note, 
          date, 
          time, 
          priority, 
          completed, 
          to_char(created_at AT TIME ZONE 'UTC', 'YYYY-MM-DD"T"HH24:MI:SS.US"Z"') as created_at_str,
          to_char(updated_at AT TIME ZONE 'UTC', 'YYYY-MM-DD"T"HH24:MI:SS.US"Z"') as updated_at_str,
          is_deleted,
          is_future_note
      FROM cluanote_tasks;
      `
    );

    remoteTasks = rows.map((r) => ({
      uuid: String(r.uuid),
      title: String(r.title),
      note: r.note ? String(r.note) : null,
      date: String(r.date),
      time: r.time ? String(r.time) : null,
      priority: String(r.priority || "medium"),
      completed: Number(r.completed || 0),
      created_at: String(r.created_at_str || new Date().toISOString()),
      updated_at: String(r.updated_at_str || new Date().toISOString()),
      is_deleted: r.is_deleted ? 1 : 0,
      is_future_note: Number(r.is_future_note || 0),
    }));
  } else {
    // Generic sync server
    const cleanUrl = url.replace(/\/+$/, "");
    const syncEndpoint = cleanUrl.endsWith("/api/sync") ? cleanUrl : `${cleanUrl}/api/sync`;
    const response = await fetch(syncEndpoint, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ local_tasks: localTasks }),
    });

    if (!response.ok) {
      const txt = await response.text();
      throw new Error(`Sync server error (${response.status}): ${txt}`);
    }

    const data = await response.json();
    if (data.pulled_tasks && Array.isArray(data.pulled_tasks)) {
      await batchUpsertFromSync(data.pulled_tasks);
      const all = await getAllTasks();
      await rescheduleAllAlarms(all);
    }

    const nowIso = new Date().toISOString();
    await AsyncStorage.setItem(LAST_SYNCED_AT_KEY, nowIso);
    return {
      success: true,
      pushedCount: data.pushed_count || 0,
      pulledCount: data.pulled_count || (data.pulled_tasks ? data.pulled_tasks.length : 0),
      pulledTasks: data.pulled_tasks,
      syncedAt: nowIso,
      message: data.message || "Synced successfully with server.",
    };
  }

  // 3. Reconcile differences: PULL newer/new remote records FIRST
  const remoteMap = new Map<string, SyncTask>(remoteTasks.map((t) => [t.uuid, t]));
  const pulledTasks: SyncTask[] = [];
  const toPush: SyncTask[] = [];

  for (const remote of remoteTasks) {
    const local = localMap.get(remote.uuid);
    if (!local) {
      // Newly created on desktop / other device -> pull!
      pulledTasks.push(remote);
    } else {
      const remoteTime = new Date(remote.updated_at).getTime();
      const localTime = new Date(local.updated_at).getTime();
      if (remoteTime > localTime) {
        pulledTasks.push(remote);
      }
    }
  }

  for (const local of localTasks) {
    const remote = remoteMap.get(local.uuid);
    if (!remote) {
      // Newly created on this device -> push!
      toPush.push(local);
    } else {
      const localTime = new Date(local.updated_at).getTime();
      const remoteTime = new Date(remote.updated_at).getTime();
      if (localTime > remoteTime) {
        toPush.push(local);
      }
    }
  }

  // 4. Merge pulled tasks into local SQLite FIRST
  if (pulledTasks.length > 0) {
    await batchUpsertFromSync(pulledTasks);
  }

  // 5. Selectively push genuinely newer local tasks to PostgreSQL
  let pushedCount = 0;
  for (const task of toPush) {
    await executeNeonSql(
      url,
      `
      INSERT INTO cluanote_tasks (
          uuid, title, note, date, time, priority, completed, created_at, updated_at, is_deleted, is_future_note
      )
      VALUES ($1, $2, $3, $4, $5, $6, $7, $8::timestamptz, $9::timestamptz, $10, $11)
      ON CONFLICT (uuid) DO UPDATE SET
          title = EXCLUDED.title,
          note = EXCLUDED.note,
          date = EXCLUDED.date,
          time = EXCLUDED.time,
          priority = EXCLUDED.priority,
          completed = EXCLUDED.completed,
          updated_at = EXCLUDED.updated_at,
          is_deleted = EXCLUDED.is_deleted,
          is_future_note = EXCLUDED.is_future_note
      WHERE EXCLUDED.updated_at >= cluanote_tasks.updated_at;
      `,
      [
        task.uuid,
        task.title,
        task.note || null,
        task.date,
        task.time || null,
        task.priority || "medium",
        task.completed,
        task.created_at,
        task.updated_at,
        task.is_deleted === 1,
        task.is_future_note ?? 0,
      ]
    );
    pushedCount++;
  }

  // Reschedule alarms if tasks changed
  if (pulledTasks.length > 0 || pushedCount > 0) {
    try {
      const all = await getAllTasks();
      await rescheduleAllAlarms(all);
    } catch (err) {
      console.warn("Failed to reschedule alarms after sync:", err);
    }
  }

  const nowIso = new Date().toISOString();
  await AsyncStorage.setItem(LAST_SYNCED_AT_KEY, nowIso);

  return {
    success: true,
    pushedCount,
    pulledCount: pulledTasks.length,
    pulledTasks,
    syncedAt: nowIso,
    message: `Sync successful. Pushed ${pushedCount} task(s), pulled ${pulledTasks.length} task(s).`,
  };
}

// ----------------------------------------------------------------------------
// Mobile Background & Foreground Auto-Sync Scheduler
// ----------------------------------------------------------------------------

let autoSyncTimer: ReturnType<typeof setTimeout> | null = null;
let isSyncInProgress = false;
let failureCount = 0;

/**
 * Triggers sync if configured and auto-sync is enabled.
 */
export async function triggerMobileSyncIfConfigured(
  onSuccess?: (result: SyncResult) => void,
  onError?: (error: string) => void
): Promise<void> {
  if (isSyncInProgress) return;
  try {
    isSyncInProgress = true;
    const config = await getSyncConfig();
    if (config.isConfigured && config.autoSync) {
      const result = await runPostgresSync();
      failureCount = 0;
      if (onSuccess) onSuccess(result);
    }
  } catch (err) {
    failureCount = Math.min(failureCount + 1, 5);
    const msg = err instanceof Error ? err.message : String(err);
    console.warn(`[CluaNote Mobile AutoSync] Sync attempt failed:`, msg);
    if (onError) onError(msg);
  } finally {
    isSyncInProgress = false;
  }
}

/**
 * Starts periodic sync and listens for app state changes (foreground resume).
 */
export function startMobileAutoSync(
  onSuccess?: (result: SyncResult) => void,
  onError?: (error: string) => void
): () => void {
  if (autoSyncTimer) {
    clearTimeout(autoSyncTimer);
  }

  const scheduleNext = (delayMs: number) => {
    autoSyncTimer = setTimeout(async () => {
      await triggerMobileSyncIfConfigured(onSuccess, onError);
      const nextDelay =
        failureCount > 0 ? Math.min(failureCount * 10 * 1000, 2 * 60 * 1000) : 30 * 1000;
      scheduleNext(nextDelay);
    }, delayMs);
  };

  // Immediate sync when app becomes active
  const subscription = AppState.addEventListener("change", (nextState: AppStateStatus) => {
    if (nextState === "active") {
      triggerMobileSyncIfConfigured(onSuccess, onError);
    }
  });

  // Initial sync attempt after 2 seconds
  scheduleNext(2000);

  return () => {
    if (autoSyncTimer) {
      clearTimeout(autoSyncTimer);
      autoSyncTimer = null;
    }
    subscription.remove();
  };
}

File: packages/rust-sync-server/src/sync.rs
use chrono::{DateTime, Utc};
use std::fs;
use url::Url;

use crate::config::{get_config_path, get_decrypted_url};
use crate::crypto::EncryptedStore;
use crate::models::{ConnectionTestResult, SyncResult, SyncTask};

pub const TABLE_INIT_SQL: &str = r#"
CREATE TABLE IF NOT EXISTS cluanote_tasks (
    uuid VARCHAR(64) PRIMARY KEY,
    title TEXT NOT NULL,
    note TEXT,
    date VARCHAR(10) NOT NULL,
    time VARCHAR(10),
    priority VARCHAR(10) NOT NULL DEFAULT 'medium',
    completed INTEGER NOT NULL DEFAULT 0,
    created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
    updated_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
    is_deleted BOOLEAN NOT NULL DEFAULT FALSE,
    is_future_note INTEGER NOT NULL DEFAULT 0
);
CREATE INDEX IF NOT EXISTS idx_cluanote_tasks_date ON cluanote_tasks(date);
CREATE INDEX IF NOT EXISTS idx_cluanote_tasks_updated ON cluanote_tasks(updated_at);
ALTER TABLE cluanote_tasks ALTER COLUMN completed TYPE INTEGER;
ALTER TABLE cluanote_tasks ADD COLUMN IF NOT EXISTS is_future_note INTEGER NOT NULL DEFAULT 0;
"#;

pub fn parse_iso_or_now(s: &str) -> DateTime<Utc> {
    if let Ok(dt) = DateTime::parse_from_rfc3339(s) {
        dt.with_timezone(&Utc)
    } else {
        Utc::now()
    }
}

pub fn map_pg_error(e: tokio_postgres::Error) -> String {
    let s = e.to_string();
    if s.contains("password authentication failed") || s.contains("authentication failed") {
        "Authentication failed: Invalid username or password.".to_string()
    } else if s.contains("database") && s.contains("does not exist") {
        "Database does not exist or access denied.".to_string()
    } else if s.contains("timed out") || s.contains("timeout") {
        "Connection timed out. Please check your host and network connection.".to_string()
    } else if s.contains("Connection refused") {
        "Connection refused by server. Verify host, port, and firewall rules.".to_string()
    } else if s.contains("SSL") || s.contains("tls") || s.contains("certificate") {
        "SSL/TLS handshake failed. Ensure your database supports SSL (e.g., sslmode=require)."
            .to_string()
    } else {
        format!("PostgreSQL Error: {}", s)
    }
}

pub fn redact_connection_url(raw: &str) -> String {
    if let Ok(mut parsed) = Url::parse(raw) {
        if parsed.password().is_some() {
            let _ = parsed.set_password(Some("••••••••"));
        }
        parsed.to_string()
    } else {
        "postgresql://••••••••".to_string()
    }
}

pub fn validate_and_sanitize_url(raw: &str) -> Result<String, String> {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return Err("Connection string cannot be empty.".to_string());
    }

    let mut parsed = Url::parse(trimmed).map_err(|e| {
        format!(
            "Invalid connection URL format. Expected 'postgresql://user:pass@host/db': {}",
            e
        )
    })?;

    let scheme = parsed.scheme();
    if scheme != "postgres" && scheme != "postgresql" {
        return Err("URL scheme must start with 'postgres://' or 'postgresql://'".to_string());
    }

    if parsed.host_str().is_none() {
        return Err("Connection URL is missing a valid host name.".to_string());
    }

    // Check for masked bullet points in password
    if let Some(pass) = parsed.password() {
        if pass.contains('•') || pass.contains("%E2%80%A2") || pass.contains("%e2%80%a2") {
            return Err("Your connection string contains masked bullet points. Please provide the actual password.".to_string());
        }
    }

    // Check for SSL requirement
    let query = parsed.query().unwrap_or("");
    if query.contains("sslmode=disable") {
        return Err("Plaintext connection (sslmode=disable) is not allowed. Please use SSL (e.g. sslmode=require) for secure synchronization.".to_string());
    }

    // Filter out unsupported channel_binding parameter for tokio-postgres TLS compatibility
    if query.contains("channel_binding") {
        let pairs: Vec<(String, String)> = parsed
            .query_pairs()
            .map(|(k, v)| (k.to_string(), v.to_string()))
            .filter(|(k, _)| k != "channel_binding")
            .collect();

        if pairs.is_empty() {
            parsed.set_query(None);
        } else {
            let mut ser = url::form_urlencoded::Serializer::new(String::new());
            for (k, v) in pairs {
                ser.append_pair(&k, &v);
            }
            parsed.set_query(Some(&ser.finish()));
        }
    }

    Ok(parsed.to_string())
}

pub async fn connect_to_postgres(url: &str) -> Result<tokio_postgres::Client, String> {
    let builder = native_tls::TlsConnector::builder();
    let tls_connector = builder
        .build()
        .map_err(|e| format!("Failed to initialize native TLS connector: {}", e))?;
    let connector = postgres_native_tls::MakeTlsConnector::new(tls_connector);

    let (client, connection) = tokio_postgres::connect(url, connector)
        .await
        .map_err(map_pg_error)?;

    tokio::spawn(async move {
        if let Err(e) = connection.await {
            eprintln!("Background Postgres connection ended: {}", e);
        }
    });

    Ok(client)
}

pub async fn test_postgres_connection_impl(
    raw_url: Option<String>,
) -> Result<ConnectionTestResult, String> {
    let url_to_use = match raw_url {
        Some(u) if !u.trim().is_empty() => u,
        _ => get_decrypted_url()?,
    };

    let sanitized_url = validate_and_sanitize_url(&url_to_use)?;
    let client = connect_to_postgres(&sanitized_url).await?;

    let row = client
        .query_one("SELECT version() as ver;", &[])
        .await
        .map_err(map_pg_error)?;

    let ver: String = row.get("ver");

    client
        .batch_execute(TABLE_INIT_SQL)
        .await
        .map_err(|e| format!("Connected, but failed to prepare 'cluanote_tasks' table: {}", map_pg_error(e)))?;

    Ok(ConnectionTestResult {
        success: true,
        message: "Successfully connected and verified 'cluanote_tasks' table.".to_string(),
        server_version: Some(ver),
    })
}

pub async fn sync_postgres_impl(local_tasks: Vec<SyncTask>) -> Result<SyncResult, String> {
    let raw_url = get_decrypted_url()?;
    let config_path = get_config_path()?;

    let bytes = fs::read(&config_path).map_err(|e| e.to_string())?;
    let mut store: EncryptedStore = serde_json::from_slice(&bytes).map_err(|e| e.to_string())?;

    let mut client = connect_to_postgres(&raw_url).await.map_err(|e| {
        let _ = update_last_error(&config_path, &mut store, &e);
        e
    })?;

    // Ensure schema exists
    client
        .batch_execute(TABLE_INIT_SQL)
        .await
        .map_err(|e| format!("Schema setup error: {}", map_pg_error(e)))?;

    let tx = client
        .transaction()
        .await
        .map_err(|e| format!("Transaction error: {}", map_pg_error(e)))?;

    // 1. Fetch all remote rows from PostgreSQL FIRST
    let rows = tx
        .query(
            r#"
            SELECT 
                uuid, 
                title, 
                note, 
                date, 
                time, 
                priority, 
                completed, 
                to_char(created_at AT TIME ZONE 'UTC', 'YYYY-MM-DD"T"HH24:MI:SS.US"Z"') as created_at_str,
                to_char(updated_at AT TIME ZONE 'UTC', 'YYYY-MM-DD"T"HH24:MI:SS.US"Z"') as updated_at_str,
                is_deleted,
                is_future_note
            FROM cluanote_tasks
            "#,
            &[],
        )
        .await
        .map_err(|e| format!("Failed to read remote tasks: {}", map_pg_error(e)))?;

    let mut remote_tasks: Vec<SyncTask> = Vec::new();
    for row in rows {
        let uuid: String = row.get("uuid");
        let title: String = row.get("title");
        let note: Option<String> = row.get("note");
        let date: String = row.get("date");
        let time: Option<String> = row.get("time");
        let priority: String = row.get("priority");
        let completed: i32 = match row.try_get::<_, i32>("completed") {
            Ok(v) => v,
            Err(_) => row.get::<_, i16>("completed") as i32,
        };
        let created_at: String = row.get("created_at_str");
        let updated_at: String = row.get("updated_at_str");
        let is_deleted_bool: bool = row.get("is_deleted");
        let is_future_note: i32 = row.try_get::<_, i32>("is_future_note").unwrap_or(0);

        remote_tasks.push(SyncTask {
            uuid,
            title,
            note,
            date,
            time,
            priority,
            completed,
            created_at,
            updated_at,
            is_deleted: if is_deleted_bool { 1 } else { 0 },
            is_future_note,
        });
    }

    // 2. Reconcile differences: PULL latest remote tasks FIRST, PUSH newer local tasks
    let local_map: std::collections::HashMap<String, &SyncTask> =
        local_tasks.iter().map(|t| (t.uuid.clone(), t)).collect();
    let remote_map: std::collections::HashMap<String, &SyncTask> =
        remote_tasks.iter().map(|t| (t.uuid.clone(), t)).collect();

    // Pull any remote task that is newer or newly added on another device
    let mut pulled_tasks: Vec<SyncTask> = Vec::new();
    for remote in &remote_tasks {
        if let Some(local) = local_map.get(&remote.uuid) {
            let remote_dt = parse_iso_or_now(&remote.updated_at);
            let local_dt = parse_iso_or_now(&local.updated_at);
            if remote_dt > local_dt {
                pulled_tasks.push(remote.clone());
            }
        } else {
            // New remote record -> pull
            pulled_tasks.push(remote.clone());
        }
    }

    // Determine which local tasks are genuinely newer and need pushing to remote
    let mut to_push: Vec<&SyncTask> = Vec::new();
    for local in &local_tasks {
        if let Some(remote) = remote_map.get(&local.uuid) {
            let local_dt = parse_iso_or_now(&local.updated_at);
            let remote_dt = parse_iso_or_now(&remote.updated_at);
            if local_dt > remote_dt {
                to_push.push(local);
            }
        } else {
            // New local record -> push
            to_push.push(local);
        }
    }

    // 3. Only push tasks that have newer changes
    let mut pushed_count = 0;
    if !to_push.is_empty() {
        let upsert_stmt = tx
            .prepare(
                r#"
                INSERT INTO cluanote_tasks (
                    uuid, title, note, date, time, priority, completed, created_at, updated_at, is_deleted, is_future_note
                )
                VALUES (
                    $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11
                )
                ON CONFLICT (uuid) DO UPDATE SET
                    title = EXCLUDED.title,
                    note = EXCLUDED.note,
                    date = EXCLUDED.date,
                    time = EXCLUDED.time,
                    priority = EXCLUDED.priority,
                    completed = EXCLUDED.completed,
                    updated_at = EXCLUDED.updated_at,
                    is_deleted = EXCLUDED.is_deleted,
                    is_future_note = EXCLUDED.is_future_note
                WHERE EXCLUDED.updated_at >= cluanote_tasks.updated_at;
                "#,
            )
            .await
            .map_err(|e| format!("Failed to prepare upsert query: {}", map_pg_error(e)))?;

        for local in to_push {
            let is_deleted_bool = local.is_deleted == 1;
            let completed_i32: i32 = local.completed as i32;
            let is_future_note_i32: i32 = local.is_future_note;
            let created_at_dt: DateTime<Utc> = parse_iso_or_now(&local.created_at);
            let updated_at_dt: DateTime<Utc> = parse_iso_or_now(&local.updated_at);

            tx.execute(
                &upsert_stmt,
                &[
                    &local.uuid,
                    &local.title,
                    &local.note,
                    &local.date,
                    &local.time,
                    &local.priority,
                    &completed_i32,
                    &created_at_dt,
                    &updated_at_dt,
                    &is_deleted_bool,
                    &is_future_note_i32,
                ],
            )
            .await
            .map_err(|e| format!("Failed to sync task '{}': {}", local.title, map_pg_error(e)))?;

            pushed_count += 1;
        }
    }

    tx.commit()
        .await
        .map_err(|e| format!("Failed to commit sync transaction: {}", map_pg_error(e)))?;

    let pulled_count = pulled_tasks.len();
    let now_iso = Utc::now().to_rfc3339();

    // Update config store with success status
    store.last_synced_at = Some(now_iso.clone());
    store.last_error = None;
    if let Ok(json_str) = serde_json::to_string_pretty(&store) {
        let _ = fs::write(&config_path, json_str);
    }

    Ok(SyncResult {
        success: true,
        pushed_count,
        pulled_count,
        pulled_tasks,
        synced_at: now_iso,
        message: format!(
            "Sync successful. Pushed {} task(s), pulled {} task(s).",
            pushed_count, pulled_count
        ),
    })
}

fn update_last_error(
    config_path: &std::path::PathBuf,
    store: &mut EncryptedStore,
    err: &str,
) -> Result<(), String> {
    store.last_error = Some(err.to_string());
    if let Ok(json_str) = serde_json::to_string_pretty(&store) {
        let _ = fs::write(config_path, json_str);
    }
    Ok(())
}
